x-pack/plugins/security/public/account_management/user_profile/user_profile.tsx

```diff
@@ -0,0 +1,580 @@
+import { FormChangesProvider, useFormChanges, useFormChangesContext } from './form_changes';
+import { FormField } from './form_field';
+import { FormLabel } from './form_label';
+import { FormRow } from './form_row';
+import { UserAvatar } from './user_avatar';
+import { UserProfileAPIClient } from './user_profile_api_client';
+import { createImageHandler, getRandomColor, IMAGE_FILE_TYPES, VALID_HEX_COLOR } from './utils';
+
+export interface UserProfileProps {
+  user: AuthenticatedUser;
+  data: {
+    avatar?: IUserAvatar;
+  };
+}
+
+export interface UserProfileFormValues {
+  user: {
+    full_name: string;
+    email: string;
+  };
+  data: {
+    avatar: {
+      initials: string;
+      color: string;
+      imageUrl: string;
+    };
+  };
+  avatarType: 'initials' | 'image';
+  customAvatarInitials: boolean;
+  customAvatarColor: boolean;
+}
+
+export const UserProfile: FunctionComponent<UserProfileProps> = ({ user, data }) => {
+  const { services } = useKibana<CoreStart>();
+  const [initialValues, resetInitialValues] = useState<UserProfileFormValues>({
+    user: {
+      full_name: user.full_name || '',
+      email: user.email || '',
+    },
+    data: {
+      avatar: {
+        initials: data.avatar?.initials || getUserAvatarInitials(user),
+        color: data.avatar?.color || getUserAvatarColor(user),
+        imageUrl: data.avatar?.imageUrl ?? '',
+      },
+    },
+    avatarType: data.avatar?.imageUrl ? 'image' : 'initials',
+    customAvatarInitials:
+      !!data.avatar?.initials && data.avatar?.initials !== getUserAvatarInitials(user),
+    customAvatarColor: !!data.avatar?.color && data.avatar?.color !== getUserAvatarColor(user),
+  });
+
+  const formik = useFormik<UserProfileFormValues>({
+    onSubmit: async (values) => {
+      try {
+        await Promise.all([
+          new UserAPIClient(services.http).saveUser({
+            username: user.username,
+            roles: user.roles,
+            enabled: user.enabled,
+            full_name: values.user.full_name,
+            email: values.user.email,
+          }),
+          new UserProfileAPIClient(services.http).update(
+            values.avatarType === 'image'
+              ? values.data
+              : { ...values.data, avatar: { ...values.data.avatar, imageUrl: null } }
+          ),
+        ]);
+        resetInitialValues(values);
+        services.notifications.toasts.addSuccess(
+          i18n.translate('xpack.spaces.management.customizeSpaceAvatar.initialsHelpText', {
+            defaultMessage: 'Profile updated',
+          })
+        );
+      } catch (error) {
+        services.notifications.toasts.addError(error, {
+          title: i18n.translate('xpack.spaces.management.customizeSpaceAvatar.initialsHelpText', {
+            defaultMessage: "Couldn't update profile",
+          }),
+        });
+      }
+    },
+    initialValues,
+    enableReinitialize: true,
+  });
+
+  const formChanges = useFormChanges();
```

Comment:
If I'm editing profile for the first time, I see `3 unsaved changes` when I just make one - ype `a` in the `Full name` field. Is it because when I change the full name, both initials and avatar color change too (it took me quite a bit to understand what has happened)?

Also when I type `Oleg`, I see `Initials` value changes to `O`, but after that `Initials` value is no longer updated even if I change the full name.

Is this intentional behavior?

Reply:
I agree this is confusing but it's one of the consequences of the proposed design. 

It might make sense for the colour field not to automatically get updated as the name changes. I think for initials it makes sense since that's based on the full name but I think colour would be less obvious to users. 

I'll have a look at the bug you mentioned.